File: services/image-service/src/middleware/metrics.py
from fastapi import Request
from prometheus_client import Counter, Histogram
from starlette.middleware.base import BaseHTTPMiddleware
import time

# Metrics
REQUEST_COUNT = Counter('http_requests_total', 'Total HTTP requests', ['method', 'endpoint', 'status'])
REQUEST_DURATION = Histogram('http_request_duration_seconds', 'HTTP request duration', ['method', 'endpoint'])

class MetricsMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        start_time = time.time()
        
        # Process request
        response = await call_next(request)
        
        # Record metrics
        duration = time.time() - start_time
        method = request.method
        path = request.url.path
        status = response.status_code
        
        REQUEST_COUNT.labels(method=method, endpoint=path, status=status).inc()
        REQUEST_DURATION.labels(method=method, endpoint=path).observe(duration)
        
        return response

File: services/auth-service/internal/middleware/metrics.go
package middleware

import (
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/prometheus/client_golang/prometheus"
)

// Metrics returns a middleware that collects HTTP metrics for Prometheus
func Metrics(requestsTotal *prometheus.CounterVec, requestDuration *prometheus.HistogramVec) gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()

		// Process request
		c.Next()

		// Collect metrics
		status := strconv.Itoa(c.Writer.Status())
		endpoint := c.FullPath()
		method := c.Request.Method
		
		// If no route was matched, use the path
		if endpoint == "" {
			endpoint = c.Request.URL.Path
		}

		// Record the request
		requestsTotal.WithLabelValues(method, endpoint, status).Inc()
		
		// Record the duration
		duration := time.Since(start)
		requestDuration.WithLabelValues(method, endpoint).Observe(duration.Seconds())
	}
}